<a href="https://colab.research.google.com/github/HazelMary/ML_TVE19CS036/blob/main/ML_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from numpy import mean,std
from matplotlib import pyplot as plt
from sklearn.model_selection import KFold
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D,MaxPooling2D,Dense,Flatten
from tensorflow.keras.optimizers import Adam

In [2]:
# load dataset
(trainX, trainy), (testX, testy) = mnist.load_data()
# summarize loaded dataset
print('Train: X=%s, y=%s' % (trainX.shape, trainy.shape))
print('Test: X=%s, y=%s' % (testX.shape, testy.shape))

11501568/11490434 [==============================] - 0s 0us/step
Train: X=(60000, 28, 28), y=(60000,)
Test: X=(10000, 28, 28), y=(10000,)


In [3]:
(trainX, trainY), (testX, testY) = mnist.load_data()

# reshape dataset to have a single channel
trainX = trainX.reshape((trainX.shape[0], 28, 28, 1))
testX = testX.reshape((testX.shape[0], 28, 28, 1))

In [4]:
from tensorflow.keras.utils import to_categorical

# one hot encode target values
trainY = to_categorical(trainY)
testY = to_categorical(testY)

In [5]:
# convert from integers to floats
train_norm = trainX.astype('float32')
test_norm = testX.astype('float32')
# normalize to range 0-1
trainX = train_norm / 255.0
testX = test_norm / 255.0

MODEL 1

In [14]:
def define_model_1():
	model = Sequential()
	model.add(Flatten(input_shape=(28, 28)))
	model.add(Dense(250, activation='relu'))
	model.add(Dense(10, activation='softmax'))
	# compile model
	opt = Adam(learning_rate=0.01)
	model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
	return model

In [15]:
# evaluate a model using k-fold cross-validation
def evaluate_model_1():
    scores, histories = list(), list()
    # define model
    model = define_model_1()
    print(model.summary())
    # fit model
    history = model.fit(trainX, trainY, epochs=10, batch_size=32, validation_data=(testX, testY), verbose=1)
    # evaluate model
    _, acc = model.evaluate(testX, testY, verbose=1)
    print('> %.3f' % (acc * 100.0))
    # stores scores
    scores.append(acc)
    histories.append(history)
    return scores, histories

In [16]:
scores, histories = evaluate_model_1()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_3 (Flatten)         (None, 784)               0         
                                                                 
 dense_8 (Dense)             (None, 250)               196250    
                                                                 
 dense_9 (Dense)             (None, 10)                2510      
                                                                 
Total params: 198,760
Trainable params: 198,760
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.2402 - accuracy: 0.9310 - val_loss: 0.1842 - val_accuracy: 0.9490
Epoch 2/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1650 - accuracy: 0.9550 - val_loss: 0.1691 - val_accuracy: 0.9548
Epoch 3/10
1875/187

MODEL 2

In [25]:
def define_model_2():
    model = Sequential()
    model.add(Flatten(input_shape=(28, 28)))
    model.add(Dense(220, activation='relu'))
    model.add(Dense(150, activation='relu'))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(10, activation='softmax'))
    # compile model
    opt = Adam(learning_rate=0.01)
    model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
    return model

In [20]:
# evaluate a model using k-fold cross-validation
def evaluate_model_2():
    scores, histories = list(), list()
    # define model
    model = define_model_2()
    print(model.summary())
    # fit model
    history = model.fit(trainX, trainY, epochs=10, batch_size=32, validation_data=(testX, testY), verbose=1)
    # evaluate model
    _, acc = model.evaluate(testX, testY, verbose=1)
    print('> %.3f' % (acc * 100.0))
    # stores scores
    scores.append(acc)
    histories.append(history)
    return scores, histories

Execute the Second Model Training

In [26]:
scores, histories = evaluate_model_2()

Model: "sequential_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_8 (Flatten)         (None, 784)               0         
                                                                 
 dense_26 (Dense)            (None, 220)               172700    
                                                                 
 dense_27 (Dense)            (None, 150)               33150     
                                                                 
 dense_28 (Dense)            (None, 64)                9664      
                                                                 
 dense_29 (Dense)            (None, 10)                650       
                                                                 
Total params: 216,164
Trainable params: 216,164
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/10
1875/1875 [=====================

# **REPORT**

Two Deep Learning Models with softmax layer activation at the top and 'relu' on other layers using Neural Network architecture was created on the MNIST dataset.

MODEL 1: \\
Number of NN layers = 2 \\
Validation Accuracy = 96.580 \\
Validation Loss = 0.1915 \\
Number of Trainable Parameters = 198,760 \\
 \\
MODEL 2: \\
Number of NN layers = 4 \\
Validation Accuracy = 96.730 \\
Validation Loss = 0.1794 \\
Number of Trainable Parameters = 216,164 \\
 \\
Model 2 has better validation accuracy than Model 1 hence Model 2 has better performance.